upup/pkg/fi/cloudup/alitasks/launchconfiguration.go

```diff
@@ -0,0 +1,272 @@
+
+	actual.ScalingGroup = &ScalingGroup{
+		ScalingGroupId: fi.String(configList[0].ScalingGroupId),
+	}
+	actual.SecurityGroup = &SecurityGroup{
+		SecurityGroupId: fi.String(configList[0].SecurityGroupId),
+	}
+
+	if len(configList[0].Tags.Tag) != 0 {
+		for _, tag := range configList[0].Tags.Tag {
+			actual.Tags[tag.Key] = tag.Value
+		}
+	}
+
+	// Ignore "system" fields
+	actual.Lifecycle = l.Lifecycle
+	return actual, nil
+}
+
+func (l *LaunchConfiguration) Run(c *fi.Context) error {
+	c.Cloud.(aliup.ALICloud).AddClusterTags(l.Tags)
+	return fi.DefaultDeltaRunMethod(l, c)
+}
+
+func (_ *LaunchConfiguration) CheckChanges(a, e, changes *LaunchConfiguration) error {
+	//Configuration can not be modified, we need to create a new one
+
+	if e.Name == nil {
+		return fi.RequiredField("Name")
+	}
+
+	if e.ImageId == nil {
+		return fi.RequiredField("ImageId")
+	}
+	if e.InstanceType == nil {
+		return fi.RequiredField("InstanceType")
+	}
+
+	return nil
+}
+
+func (_ *LaunchConfiguration) RenderALI(t *aliup.ALIAPITarget, a, e, changes *LaunchConfiguration) error {
+
+	glog.V(2).Infof("Creating LaunchConfiguration for ScalingGroup:%q", fi.StringValue(e.ScalingGroup.ScalingGroupId))
+
+	createScalingConfiguration := &ess.CreateScalingConfigurationArgs{
+		ScalingGroupId:      fi.StringValue(e.ScalingGroup.ScalingGroupId),
+		ImageId:             fi.StringValue(e.ImageId),
+		InstanceType:        fi.StringValue(e.InstanceType),
+		SecurityGroupId:     fi.StringValue(e.SecurityGroup.SecurityGroupId),
+		SystemDisk_Size:     common.UnderlineString(strconv.Itoa(fi.IntValue(e.SystemDiskSize))),
+		SystemDisk_Category: common.UnderlineString(fi.StringValue(e.SystemDiskCategory)),
+	}
+
+	if e.RAMRole != nil && e.RAMRole.Name != nil {
+		createScalingConfiguration.RamRoleName = fi.StringValue(e.RAMRole.Name)
+	}
+
+	if e.UserData != nil {
+		userData, err := e.UserData.AsString()
+		if err != nil {
+			return fmt.Errorf("error rendering ScalingLaunchConfiguration UserData: %v", err)
+		}
+		createScalingConfiguration.UserData = userData
+	}
+
+	if e.SSHKey != nil && e.SSHKey.Name != nil {
+		createScalingConfiguration.KeyPairName = fi.StringValue(e.SSHKey.Name)
+	}
+
+	if e.Tags != nil {
+		tagItem, err := json.Marshal(e.Tags)
+		if err != nil {
+			return fmt.Errorf("error rendering ScalingLaunchConfiguration Tags: %v", err)
+		}
+		createScalingConfiguration.Tags = string(tagItem)
+	}
+
+	createScalingConfigurationResponse, err := t.Cloud.EssClient().CreateScalingConfiguration(createScalingConfiguration)
+	if err != nil {
+		return fmt.Errorf("error creating scalingConfiguration: %v", err)
+	}
+	e.ConfigurationId = fi.String(createScalingConfigurationResponse.ScalingConfigurationId)
+
+	// Disable ScalingGroup, used to bind scalingConfig, we should excute EnableScalingGroup in the task LaunchConfiguration
+	// If the ScalingGroup is active, we can not excute EnableScalingGroup.
+	if e.ScalingGroup.Active != nil && fi.BoolValue(e.ScalingGroup.Active) {
+
```

Comment:
Should we be looking here at `a`, not `e`?  I think we're disabling the existing scaling group?

Reply:
We will not create a new scaling group when 'update', we just create a new configuration for the old scaling group, so both 'a' and 'e' will have the correct id of exiting scaling group. But use 'a' will be more reasonable here, I will remember this tip. Thank you very much for your instant review of the codes.